### DISCUSSION TITLE
Noise Protocol Framework Integration

### DISCUSSION THREAD

**Discussion by centromere** [capnproto/capnproto#1850] [Ideas]
I am starting this thread to discuss the integration of the [Noise Protocol Framework](https://noiseprotocol.org) into kj/capnp, which will serve as the basis for level 3 protocol features.

This discussion originated [here](https://github.com/capnproto/capnproto/pull/1845#issuecomment-1821609366).

The opening questions are:
* How are keys managed and shared?
* Is zero-round-trip startup supported? If so, how?
* Does the design consider the future of three-party handoff (which is not implemented yet), or is it specifically for two-party connections?

> How are keys managed and shared?

I think it would be best to leave key management up to the application. In other words, the application should responsible for (de)serializing long term keys and for performing more comprehensive authentication checks, if desired. The RPC protocol will provide a standard way to serialize keys, and this data structure will be used when sharing keys with other vats (such as during a third-party hand-off).

> Is zero-round-trip startup supported? If so, how?

Yes. It is [explicitly supported](https://noiseprotocol.org/noise.html#zero-rtt-and-noise-protocols) by Noise. Note however that a suitable handshake must be chosen:
> Patterns where the initiator has pre-knowledge of the responder's static public key (i.e. patterns ending in K) allow zero-RTT encryption, meaning the initiator can encrypt the first handshake payload.

My intention to ensure that the C++ API allows for this, even if it's not immediately implemented.

> Does the design consider the future of three-party handoff (which is not implemented yet), or is it specifically for two-party connections?

My intention is to enable support for three-party handoff from the outset, even if it's not implemented immediately.

What follows is my first attempt at describing the data structures present in the RPC protocol:

```capnp
@0xe7174c4e92d15add;

using Cxx = import "/capnp/c++.capnp";
$Cxx.namespace("capnp::rpc::noise");

struct Bytes16 {
  a @0 :UInt64;
  b @1 :UInt64;
}

struct Bytes32 {
  a @0 :UInt64;
  b @1 :UInt64;
  c @2 :UInt64;
  d @3 :UInt64;
}

struct Bytes56 {
  a @0 :UInt64;
  b @1 :UInt64;
  c @2 :UInt64;
  d @3 :UInt64;
  e @4 :UInt64;
  f @5 :UInt64;
  g @6 :UInt64;
}

struct PublicKey {
  union {
    x25519 :group {
      bytes @0 :Bytes32;
    }

    x448 :group {
      bytes @1 :Bytes56;
    }
  }
}

struct VatId {
  publicKey @0 :PublicKey;
}

using Nonce = UInt64;

struct ProvisionId {
  providerVatId @0 :VatId;
  nonce @1 :Nonce;
}

struct RecipientId {
  recipientVatId @0 :VatId;
  nonce @1 :Nonce;
}

struct VatAddress {
  union {
    ipv6 :group {
      bytes @0 :Bytes16;
    }

    ipv4 :group {
      bytes @1 :UInt32;
    }
  }
}

struct ThirdPartyCapId {
  address @0 :VatAddress;
  vatId @1 :VatId;
}

struct JoinResult {}
```

**Comment by kentonv** [comment]
Sorry for slow reply here. This has been in my inbox but every time I look at it I think "Well, I will need to review this carefully and I don't have time right now"... :/

When I last looked into zero-RTT and Noise (a long time ago!) there was some tension about "perfect forward secrecy" (PFS): Normally, you want each side to generate an ephemeral keypair per-connection, and use those to generate the symmetric key. Long-term keys are used only to sign the short-term keys. But this implies that the client _must_ wait for the server to tell it the ephemeral public key it has chosen for the connection, before the client can then derive the symmetric secret and start encrypting things.

If we take ephemeral keys out of the mix and just derive a shared secret based on each side's permanent key, then we have the problem that if either key is ever leaked, a previously-recorded session can be decrypted. Ephemeral keys make this much less likely since the private key typically is never stored to disk nor even leaves the memory address space of the owner.

In Cap'n Proto, though, with three-party handoff, there are a couple compromises to consider. If Alice is connected to Bob and Carol, and Alice tells Bob to directly connect to Carol, Alice can help facilitate a 0-RTT connection in a way that introduces ephemeral key material, thus not breaking PFS. There are actually two possibilities I can think of:

1. Alice could take Carol's ephemeral key -- which was established earlier -- and pass it along to Bob. Bob could use this ephemeral key generate a symmetric key for talking to Carol. This might be used just for the first few messages until Carol sends back a new key to use, or maybe they just go ahead and keep using that key. In one version of this idea, each _running process_ generates a single ephemeral key pair at startup and uses that for all connections it participates in. This still has the property that the private key is never stored nor transmitted, although it's true the key sticks around in memory longer than a single connection.

2. Alice could simply generate a symmetric key for Bob and Carol to use, and send it to both of them. No asymmetric keys are needed at all. This would certainly be a lot simpler and faster! You might say "oh no, Alice could maliciously use the key to spy on the connection between Bob and Carol". But, this is not a real problem: Alice is the one who told Bob and Carol to talk in the first place, and could just as easily have told them both to talk through a proxy controlled by Alice. The real problem comes when a fourth party, Dave, _also_ tries to introduce Bob and Carol. Now Bob and Carol end up with two different connections, one created in response to Alice, and one in response to Dave. Maybe they could use some complicated protocol to merge these connections safely. However, everything would be easier using option (1) if Carol used the same public key for every connection, and therefore Alice and Dave both passed along the same key to Bob. Then Bob knows he can safely reuse the same connection.

I'm sort of conflicted between these two approaches. I guess (1) seems better in the case that each of the parties really is a single Vat (a single process / event loop). It gets weird when you start thinking about distributed systems where connections are load balanced, and therefore when Alice introduces Bob to Carol, Bob might end up talking to a different instance of Carol than the one Alice was talking to. In this case, either all instances of Carol need to have the same keypair (which means it is being transmitted, and therefore more likely to leak?), or there would need to be some way to route to the _same_ instance when needed... Approach (2) has a different problem in this case, which is that whichever instance of Carol that Bob ends up connecting to needs to somehow get ahold of the symmetric key from the instance Alice had connected to... but that might be easier? Dunno.

**Reply by centromere** [reply]
I appreciate your thoughtful response, Kenton. Have you considered the `KK` pattern?

```
KK:
     -> s
     <- s
     ...
     -> e, es, ss
     <- e, ee, se
```

From your example,
> Alice is connected to Bob and Carol, and Alice tells Bob to directly connect to Carol.

Alice can send the following messages:

```
Alice -> Bob: ThirdPartyCapId(Carol)
Alice -> Carol: ThirdPartyCapId(Bob)
```

In my proposal, the `VatId` within the `ThirdPartyCapId` *is* the `s` value from the Noise protocol. Bob can use the `VatAddress` contained with the `ThirdPartyCapId` to establish a streaming connection to Carol. The identity of both Bob and Carol are known to one another, and the `KK` pattern can proceed accordingly. Bob can send application data in his initial message, thus fulfilling the 0-RTT requirement.
